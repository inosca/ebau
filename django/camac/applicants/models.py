from django.db import models


class ROLE_CHOICES(models.TextChoices):
    ADMIN = "ADMIN", "Admin"
    EDITOR = "EDITOR", "Editor"
    READ_ONLY = "READ_ONLY", "Read only"


class Applicant(models.Model):
    instance = models.ForeignKey(
        "instance.Instance",
        models.CASCADE,
        db_column="INSTANCE_ID",
        # NOTE: The "involved_" prefix is required because the instance views annotate
        # "applicants". And we did not wanted to break backwards compatibility
        related_name="involved_applicants",
    )
    user = models.ForeignKey(
        "user.User", models.DO_NOTHING, db_column="USER_ID", related_name="+"
    )
    invitee = models.ForeignKey(
        "user.User",
        models.DO_NOTHING,
        db_column="APPLICANT_USER_ID",
        related_name="+",
        blank=True,
        null=True,
    )
    created = models.DateTimeField(db_column="CREATED", auto_now=True)
    email = models.EmailField(db_collation="case_insensitive")
    role = models.CharField(
        max_length=50,
        choices=ROLE_CHOICES.choices,
        default=ROLE_CHOICES.ADMIN.value,
    )

    class Meta:
        managed = True
        db_table = "APPLICANTS"
        unique_together = (("instance", "invitee"),)
