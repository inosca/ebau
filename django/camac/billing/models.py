from django.db import models


class BillingV2Entry(models.Model):
    TAX_MODE_INCLUSIVE = "inclusive"
    TAX_MODE_EXCLUSIVE = "exclusive"
    TAX_MODE_EXEMPT = "exempt"
    TAX_MODE_CHOICES = (
        (TAX_MODE_INCLUSIVE, "Incl 7.7%"),
        (TAX_MODE_EXCLUSIVE, "Excl 7.7%"),
        (TAX_MODE_EXEMPT, "Tax exempt"),
    )

    CALCULATION_FLAT = "flat"
    CALCULATION_PERCENTAGE = "percentage"
    CALCULATION_HOURLY = "hourly"
    CALCULATION_CHOICES = (
        (CALCULATION_FLAT, "Flat rate"),
        (CALCULATION_PERCENTAGE, "Percentage"),
        (CALCULATION_HOURLY, "Hourly"),
    )

    MUNICIPAL = "municipal"
    CANTONAL = "cantonal"
    ORGANIZATION_CHOICES = ((MUNICIPAL, "Municipal"), (CANTONAL, "Cantonal"))

    DECIMAL_FORMAT = {
        "max_digits": 10,
        "decimal_places": 2,
        "null": True,
        "blank": True,
    }

    # Structural: Which instance is the item billed to?
    instance = models.ForeignKey("instance.Instance", models.CASCADE, related_name="+")

    # Organisation: Who charged the item?
    group = models.ForeignKey("user.Group", models.DO_NOTHING, related_name="+")
    user = models.ForeignKey("user.User", models.DO_NOTHING, related_name="+")

    # Billing text
    text = models.TextField()
    date_added = models.DateField(auto_now_add=True)
    date_charged = models.DateField(null=True, blank=True)

    # Tax mode = calculation model for tax
    tax_mode = models.CharField(choices=TAX_MODE_CHOICES, max_length=20)

    # Calculation mode
    calculation = models.CharField(choices=CALCULATION_CHOICES, max_length=20)

    # Tax rate (percentage)
    tax_rate = models.DecimalField(**DECIMAL_FORMAT)

    # Calculation mode: hourly rate
    hours = models.DecimalField(**DECIMAL_FORMAT)
    hourly_rate = models.DecimalField(**DECIMAL_FORMAT)

    # Calculation mode: percentage of total cost
    percentage = models.DecimalField(**DECIMAL_FORMAT)
    total_cost = models.DecimalField(**DECIMAL_FORMAT)

    # Final rate (may be entered directly in "flat" mode, otherwise it's
    # calculated. We store it for easier handling in the output however.
    final_rate = models.DecimalField(**DECIMAL_FORMAT)

    # Organization: either municipal or cantonal but can be NULL
    # Used to distinguish which oranization collects part of the bill
    organization = models.CharField(
        choices=ORGANIZATION_CHOICES, max_length=20, null=True
    )
