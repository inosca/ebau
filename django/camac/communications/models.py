import os.path

from django.contrib.postgres.fields import ArrayField
from django.db import models

from camac.user.permissions import permission_aware


def entity_for_current_user(request):
    """Return Entity string for the current user.

    This is the string "APPLICANT" if the user
    is an applicant (Ie. has no service groups)
    or the currently-relevant Service ID (also as string)
    """
    return EntityInfo(request).entity() or ""


class EntityInfo:
    def __init__(self, request):
        self.request = request
        self._resolved = None

    @permission_aware
    def entity(self):
        # Return the service ID of the default group, if the default group
        # actually has a service ID. Otherwise, fall back to the first group
        # of the user that does have a service ID.
        # Finally, fall back to empty string if that's also not the case.
        return str(
            self.request.group.service_id
            or self.request.user.get_default_group().service_id
            or (
                self.request.user.groups.filter(service__isnull=False)
                .first()
                .service_id
            )
        )

    def entity_for_applicant(self):
        return "APPLICANT"


class CommunicationsTopic(models.Model):
    instance = models.ForeignKey(
        "instance.Instance",
        on_delete=models.CASCADE,
    )
    initiated_by = models.ForeignKey(
        "user.User", on_delete=models.DO_NOTHING, related_name="+"
    )
    subject = models.TextField()
    created = models.DateTimeField(auto_now_add=True)

    allow_replies = models.BooleanField(default=True)

    involved_entities = ArrayField(models.CharField(max_length=50), blank=True)
    initiated_by_entity = models.CharField(max_length=50)


class CommunicationsReadMarker(models.Model):
    read_at = models.DateTimeField(auto_now_add=True)
    message = models.ForeignKey(
        "CommunicationsMessage", on_delete=models.CASCADE, related_name="read_by"
    )
    entity = models.CharField(max_length=50)


class CommunicationsMessage(models.Model):
    topic = models.ForeignKey(
        CommunicationsTopic, on_delete=models.CASCADE, related_name="messages"
    )
    body = models.TextField()

    created_by = models.CharField(max_length=50)
    created_by_user = models.ForeignKey(
        "user.User", on_delete=models.DO_NOTHING, related_name="+", null=True
    )
    created_at = models.DateTimeField(auto_now_add=True)
    sent_at = models.DateTimeField(default=None, null=True, blank=True)

    def mark_as_read_by_entity(self, entity):
        """Mark this message, and all messages before, as "read".

        When a message is marked as "read", we assume the user has also
        seen all messages in the same topic that were there before it.
        """
        # created_at__lte=self.created_at includes self, so no
        # explicit marking for "self"
        for message in self.topic.messages.filter(created_at__lte=self.created_at):
            message.read_by.get_or_create(entity=entity)


def attachment_path_directory_path(attachment, filename):
    return "communications/files/{0}/{1}/{2}".format(
        attachment.message.topic.instance_id, attachment.message.topic.pk, filename
    )


class CommunicationsAttachment(models.Model):
    message = models.ForeignKey(
        CommunicationsMessage, on_delete=models.CASCADE, related_name="attachments"
    )
    file_attachment = models.FileField(
        null=True, default=None, blank=True, upload_to=attachment_path_directory_path
    )
    file_type = models.CharField(max_length=250, null=True, default=None, blank=True)

    document_attachment = models.ForeignKey(
        "document.Attachment",
        on_delete=models.DO_NOTHING,
        related_name="+",
        blank=True,
        null=True,
        default=None,
    )

    def __str__(self):
        atype = "uploaded" if self.file_attachment else "via docs module"
        msg = self.message_id and str(self.message) or "(no msg)"
        return f"Attachment on {msg}: {self.filename} ({atype})"

    @property
    def filename(self):
        if self.file_attachment:
            return os.path.basename(self.file_attachment.name)
        return os.path.basename(self.document_attachment.name)
