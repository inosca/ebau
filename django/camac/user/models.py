import hashlib

from django.conf import settings
from django.contrib.auth.base_user import AbstractBaseUser, BaseUserManager
from django.contrib.postgres.fields import CIEmailField
from django.db import models
from django.utils.translation import gettext as _

from ..core import models as core_models


class UserManager(BaseUserManager):
    use_in_migrations = True

    def create_user(self, username, email=None, password=None, **extra_fields):
        email = self.normalize_email(email)
        user = self.model(username=username, email=email, **extra_fields)
        user.set_password(password)
        user.save(using=self._db)
        return user

    def create_superuser(self, username, email, password, **extra_fields):
        return self.create_user(username, email, password, **extra_fields)


class User(AbstractBaseUser):
    objects = UserManager()

    USERNAME_FIELD = "username"
    REQUIRED_FIELDS = ["email", "name", "surname", "language"]

    id = models.AutoField(db_column="USER_ID", primary_key=True, verbose_name=_("ID"))
    username = models.CharField(
        db_column="USERNAME", unique=True, max_length=250, verbose_name=_("Username")
    )
    password = models.CharField(
        db_column="PASSWORD", max_length=50, blank=True, null=True
    )
    name = models.CharField(
        db_column="NAME", max_length=100, verbose_name=_("First name")
    )
    surname = models.CharField(
        db_column="SURNAME", max_length=100, verbose_name=_("Last name")
    )
    email = CIEmailField(
        db_column="EMAIL",
        max_length=100,
        blank=True,
        null=True,
        verbose_name=_("Email"),
    )
    phone = models.CharField(
        db_column="PHONE",
        max_length=100,
        blank=True,
        null=True,
        verbose_name=_("Phone"),
    )
    disabled = models.PositiveSmallIntegerField(
        db_column="DISABLED", default=0, verbose_name=_("Disabled?")
    )
    language = models.CharField(
        db_column="LANGUAGE", max_length=2, verbose_name=_("Language")
    )
    last_login = models.DateTimeField(
        db_column="LAST_REQUEST_DATE",
        blank=True,
        null=True,
        verbose_name=_("Last login"),
    )
    address = models.CharField(
        db_column="ADDRESS",
        max_length=100,
        blank=True,
        null=True,
        verbose_name=_("Address"),
    )
    city = models.CharField(
        db_column="CITY", max_length=100, blank=True, null=True, verbose_name=_("City")
    )
    zip = models.CharField(
        db_column="ZIP", max_length=10, blank=True, null=True, verbose_name=_("Zip")
    )
    groups = models.ManyToManyField("Group", through="UserGroup", related_name="users")

    @property
    def is_superuser(self):
        return UserGroup.objects.filter(
            user=self, group_id=settings.APPLICATION.get("ADMIN_GROUP")
        ).exists()  # pragma: no cover

    @property
    def is_staff(self):
        return self.is_superuser  # pragma: no cover

    def has_perm(self, perm):
        """
        Since we don't use the PermissionsMixin this method has to be overwritten.

        Intially this method returns True if the user has the specified permission.
        We check this in the is_superuser property.
        """
        return True  # pragma: no cover

    def has_module_perms(self, app_label):
        """
        Since we don't use the PermissionsMixin this method has to be overwritten.

        Intially this method returns True if the user has any permission in the given package.
        We check this in the is_superuser property.
        """
        return True  # pragma: no cover

    def _make_password(self, raw_password):
        salted = settings.AUTH_PASSWORT_SALT + raw_password
        return hashlib.md5(salted.encode()).hexdigest()

    def set_password(self, raw_password):
        self.password = self._make_password(raw_password)
        self._password = raw_password

    def check_password(self, raw_password):
        """
        Check password whether it matches.

        Empty passwords are not allowed.
        """
        password_empty = raw_password in ["", None]

        return not password_empty and self.password == self._make_password(raw_password)

    def get_full_name(self):
        return "{0} {1}".format(self.name, self.surname).strip()

    @property
    def is_active(self):
        return not self.disabled

    def get_default_group(self):
        user_group = UserGroup.objects.filter(user=self, default_group=True).first()
        if user_group:
            return user_group.group

    def __str__(self):
        return self.get_full_name()

    class Meta:
        managed = True
        db_table = "USER"
        verbose_name = _("User")
        verbose_name_plural = _("Users")


class UserT(models.Model):
    user = models.ForeignKey(
        User, models.DO_NOTHING, db_column="USER_ID", related_name="+"
    )
    language = models.CharField(db_column="LANGUAGE", max_length=2)
    city = models.CharField(db_column="CITY", max_length=100, blank=True, null=True)

    class Meta:
        managed = True
        db_table = "USER_T"


class Group(core_models.MultilingualModel, models.Model):
    group_id = models.AutoField(
        db_column="GROUP_ID", primary_key=True, verbose_name=_("ID")
    )
    role = models.ForeignKey(
        "user.Role",
        models.PROTECT,
        db_column="ROLE_ID",
        related_name="groups",
        verbose_name=_("Role"),
    )
    service = models.ForeignKey(
        "user.Service",
        models.SET_NULL,
        db_column="SERVICE_ID",
        related_name="groups",
        blank=True,
        null=True,
        verbose_name=_("Service"),
    )
    disabled = models.PositiveSmallIntegerField(
        db_column="DISABLED", default=0, verbose_name=_("Disabled?")
    )
    name = models.CharField(
        db_column="NAME", max_length=100, blank=True, null=True, verbose_name=_("Name")
    )
    phone = models.CharField(
        db_column="PHONE",
        max_length=100,
        blank=True,
        null=True,
        verbose_name=_("Phone"),
    )
    zip = models.CharField(
        db_column="ZIP", max_length=10, blank=True, null=True, verbose_name=_("Zip")
    )
    city = models.CharField(
        db_column="CITY", max_length=100, blank=True, null=True, verbose_name=_("City")
    )
    address = models.CharField(
        db_column="ADDRESS",
        max_length=100,
        blank=True,
        null=True,
        verbose_name=_("Address"),
    )
    email = models.CharField(
        db_column="EMAIL",
        max_length=100,
        blank=True,
        null=True,
        verbose_name=_("Email"),
    )
    website = models.CharField(
        db_column="WEBSITE",
        max_length=1000,
        blank=True,
        null=True,
        verbose_name=_("Website"),
    )
    locations = models.ManyToManyField(
        "Location", through="GroupLocation", verbose_name=_("Locations")
    )

    class Meta:
        managed = True
        db_table = "GROUP"
        verbose_name = _("Group")
        verbose_name_plural = _("Groups")


class GroupT(models.Model):
    group = models.ForeignKey(
        Group,
        models.CASCADE,
        db_column="GROUP_ID",
        related_name="trans",
        verbose_name=_("Group"),
    )
    language = models.CharField(
        db_column="LANGUAGE", max_length=2, verbose_name=_("Language")
    )
    name = models.CharField(
        db_column="NAME", max_length=200, blank=True, null=True, verbose_name=_("Name")
    )
    city = models.CharField(
        db_column="CITY", max_length=100, blank=True, null=True, verbose_name=_("City")
    )

    class Meta:
        managed = True
        db_table = "GROUP_T"


class Location(core_models.MultilingualModel, models.Model):
    location_id = models.AutoField(db_column="LOCATION_ID", primary_key=True)
    communal_cantonal_number = models.IntegerField(
        db_column="COMMUNAL_CANTONAL_NUMBER", blank=True, null=True
    )
    communal_federal_number = models.CharField(
        db_column="COMMUNAL_FEDERAL_NUMBER", max_length=255, blank=True, null=True
    )
    district_number = models.IntegerField(
        db_column="DISTRICT_NUMBER", blank=True, null=True
    )
    section_number = models.IntegerField(
        db_column="SECTION_NUMBER", blank=True, null=True
    )
    name = models.CharField(db_column="NAME", max_length=100, blank=True, null=True)
    commune_name = models.CharField(
        db_column="COMMUNE_NAME", max_length=100, blank=True, null=True
    )
    district_name = models.CharField(
        db_column="DISTRICT_NAME", max_length=100, blank=True, null=True
    )
    section_name = models.CharField(
        db_column="SECTION_NAME", max_length=100, blank=True, null=True
    )
    zip = models.CharField(db_column="ZIP", max_length=10, blank=True, null=True)

    class Meta:
        managed = True
        db_table = "LOCATION"


class LocationT(models.Model):
    location = models.ForeignKey(
        Location, models.CASCADE, db_column="LOCATION_ID", related_name="trans"
    )
    language = models.CharField(db_column="LANGUAGE", max_length=2)
    name = models.CharField(db_column="NAME", max_length=100, blank=True, null=True)
    commune_name = models.CharField(
        db_column="COMMUNE_NAME", max_length=100, blank=True, null=True
    )
    district_name = models.CharField(
        db_column="DISTRICT_NAME", max_length=100, blank=True, null=True
    )
    section_name = models.CharField(
        db_column="SECTION_NAME", max_length=100, blank=True, null=True
    )

    class Meta:
        managed = True
        db_table = "LOCATION_T"


class GroupLocation(models.Model):
    id = models.AutoField(db_column="ID", primary_key=True)
    group = models.ForeignKey(
        Group,
        models.CASCADE,
        db_column="GROUP_ID",
        related_name="+",
        verbose_name=_("Group"),
    )
    location = models.ForeignKey(
        Location,
        models.CASCADE,
        db_column="LOCATION_ID",
        related_name="+",
        verbose_name=_("Location"),
    )

    class Meta:
        managed = True
        db_table = "GROUP_LOCATION"
        unique_together = (("group", "location"),)


class UserGroup(models.Model):
    id = models.AutoField(db_column="ID", primary_key=True)
    user = models.ForeignKey(
        User,
        models.CASCADE,
        db_column="USER_ID",
        related_name="user_groups",
        verbose_name=_("User"),
    )
    group = models.ForeignKey(
        Group,
        models.CASCADE,
        db_column="GROUP_ID",
        related_name="+",
        verbose_name=_("Group"),
    )
    default_group = models.PositiveSmallIntegerField(
        db_column="DEFAULT_GROUP", verbose_name=_("Default group?")
    )

    class Meta:
        managed = True
        db_table = "USER_GROUP"
        unique_together = (("user", "group"),)


class UserGroupLog(models.Model):
    user_group_log_id = models.AutoField(
        db_column="USER_GROUP_LOG_ID", primary_key=True
    )
    modification_date = models.DateTimeField(db_column="MODIFICATION_DATE")
    user_id = models.IntegerField(db_column="USER_ID")
    action = models.CharField(db_column="ACTION", max_length=500)
    data = models.TextField(db_column="DATA", blank=True, null=True)
    id1 = models.IntegerField(db_column="ID1")
    field1 = models.CharField(db_column="FIELD1", max_length=30)
    id2 = models.IntegerField(db_column="ID2")
    field2 = models.CharField(db_column="FIELD2", max_length=30)

    class Meta:
        managed = True
        db_table = "USER_GROUP_LOG"


class Role(core_models.MultilingualModel, models.Model):
    role_id = models.AutoField(db_column="ROLE_ID", primary_key=True)
    role_parent = models.ForeignKey(
        "self",
        models.SET_NULL,
        db_column="ROLE_PARENT_ID",
        related_name="+",
        blank=True,
        null=True,
    )
    name = models.CharField(db_column="NAME", max_length=100, blank=True, null=True)
    group_prefix = models.CharField(
        db_column="GROUP_PREFIX", max_length=100, blank=True, null=True
    )

    class Meta:
        managed = True
        db_table = "ROLE"


class RoleT(models.Model):
    role = models.ForeignKey(
        Role, models.CASCADE, db_column="ROLE_ID", related_name="trans"
    )
    language = models.CharField(db_column="LANGUAGE", max_length=2)
    name = models.CharField(db_column="NAME", max_length=100, blank=True, null=True)
    group_prefix = models.CharField(
        db_column="GROUP_PREFIX", max_length=100, blank=True, null=True
    )

    class Meta:
        managed = True
        db_table = "ROLE_T"


class ServiceGroup(core_models.MultilingualModel, models.Model):
    service_group_id = models.AutoField(db_column="SERVICE_GROUP_ID", primary_key=True)
    name = models.CharField(db_column="NAME", max_length=100, blank=True, null=True)
    sort = models.IntegerField(blank=True, null=True)

    class Meta:
        managed = True
        db_table = "SERVICE_GROUP"


class ServiceGroupT(models.Model):
    service_group = models.ForeignKey(
        ServiceGroup, models.CASCADE, db_column="SERVICE_GROUP_ID", related_name="trans"
    )
    language = models.CharField(db_column="LANGUAGE", max_length=2)
    name = models.CharField(db_column="NAME", max_length=100, blank=True, null=True)

    class Meta:
        managed = True
        db_table = "SERVICE_GROUP_T"


class Service(core_models.MultilingualModel, models.Model):
    service_id = models.AutoField(
        db_column="SERVICE_ID", primary_key=True, verbose_name=_("ID")
    )
    service_group = models.ForeignKey(
        ServiceGroup,
        models.PROTECT,
        db_column="SERVICE_GROUP_ID",
        related_name="+",
        verbose_name=_("Service group"),
    )
    service_parent = models.ForeignKey(
        "self",
        models.SET_NULL,
        db_column="SERVICE_PARENT_ID",
        related_name="+",
        blank=True,
        null=True,
        verbose_name=_("Service parent"),
    )
    name = models.CharField(
        db_column="NAME", max_length=100, blank=True, null=True, verbose_name=_("Name")
    )
    description = models.CharField(
        db_column="DESCRIPTION",
        max_length=255,
        blank=True,
        null=True,
        verbose_name=_("Description"),
    )
    sort = models.IntegerField(db_column="SORT")
    phone = models.CharField(
        db_column="PHONE",
        max_length=100,
        blank=True,
        null=True,
        verbose_name=_("Phone"),
    )
    zip = models.CharField(
        db_column="ZIP", max_length=10, blank=True, null=True, verbose_name=_("Zip")
    )
    city = models.CharField(
        db_column="CITY", max_length=100, blank=True, null=True, verbose_name=_("City")
    )
    address = models.CharField(
        db_column="ADDRESS",
        max_length=100,
        blank=True,
        null=True,
        verbose_name=_("Address"),
    )
    email = models.CharField(
        db_column="EMAIL",
        max_length=1000,
        blank=True,
        null=True,
        verbose_name=_("Email"),
    )
    website = models.CharField(
        db_column="WEBSITE",
        max_length=1000,
        blank=True,
        null=True,
        verbose_name=_("Website"),
    )
    disabled = models.PositiveSmallIntegerField(
        db_column="DISABLED", default=0, verbose_name=_("Disabled?")
    )
    notification = models.PositiveSmallIntegerField(
        default=1, verbose_name=_("Receive notifications?")
    )

    class Meta:
        managed = True
        db_table = "SERVICE"
        ordering = ["service_group__name"]
        verbose_name = _("Service")
        verbose_name_plural = _("Services")


class ServiceT(models.Model):
    service = models.ForeignKey(
        Service, models.CASCADE, db_column="SERVICE_ID", related_name="trans"
    )
    language = models.CharField(
        db_column="LANGUAGE", max_length=2, verbose_name=_("Language")
    )
    name = models.CharField(
        db_column="NAME", max_length=200, blank=True, null=True, verbose_name=_("Name")
    )
    description = models.CharField(
        db_column="DESCRIPTION",
        max_length=255,
        blank=True,
        null=True,
        verbose_name=_("Description"),
    )
    city = models.CharField(
        db_column="CITY", max_length=100, blank=True, null=True, verbose_name=_("City")
    )

    class Meta:
        managed = True
        db_table = "SERVICE_T"
