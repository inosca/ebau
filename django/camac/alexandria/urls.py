from alexandria.core import views as alexandria_views
from django.urls import path
from rest_framework.routers import SimpleRouter

from camac.alexandria import views

r = SimpleRouter(trailing_slash=False)

# override the default viewset with our patched versions
# this needs to be updated when new urls are added in alexandria
r.register(r"categories", views.PatchedCategoryViewSet)
r.register(r"documents", views.PatchedDocumentViewSet)
r.register(r"files", views.PatchedFileViewSet)
r.register(r"tags", views.PatchedTagViewSet)
r.register(r"tagsynonymgroups", views.PatchedTagSynonymGroupViewSet)
r.register(r"marks", views.PatchedMarkViewSet)

urlpatterns = r.urls

urlpatterns.append(path("hook", alexandria_views.hook_view, name="hook"))
