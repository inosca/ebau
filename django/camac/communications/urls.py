from rest_framework.routers import SimpleRouter

from . import views

r = SimpleRouter(trailing_slash=False)

r.register(r"communications-topics", views.TopicView, "communications-topic")
r.register(r"communications-messages", views.MessageView, "communications-message")
r.register(
    r"communications-attachments", views.AttachmentView, "communications-attachment"
)

urlpatterns = r.urls
