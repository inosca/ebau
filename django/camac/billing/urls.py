from django.urls import re_path
from rest_framework.routers import SimpleRouter

from camac.billing import views
from camac.billing.export.views import BillingV2EntryExportView

urlpatterns = [
    re_path(
        r"billing-v2-entries/export$",
        BillingV2EntryExportView.as_view(),
        name="billing-export",
    ),
]

r = SimpleRouter(trailing_slash=False)

r.register(r"billing-v2-entries", views.BillingV2EntryViewset, "billing-v2-entry")

urlpatterns.extend(r.urls)
