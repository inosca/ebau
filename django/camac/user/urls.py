from django.urls import re_path
from rest_framework.routers import SimpleRouter

from . import views

r = SimpleRouter(trailing_slash=False)

r.register(r"locations", views.LocationView)
r.register(r"groups", views.GroupView)
r.register(r"public-groups", views.PublicGroupView, basename="publicgroup")
r.register(r"roles", views.RoleView)
r.register(r"services", views.ServiceView)
r.register(r"public-services", views.PublicServiceView, basename="publicservice")
r.register(r"users", views.UserView)
r.register(r"public-users", views.PublicUserView, basename="publicuser")
r.register(r"user-groups", views.UserGroupView, basename="usergroup")

urlpatterns = [re_path(r"^me", views.MeView.as_view({"get": "retrieve"}), name="me")]

urlpatterns.extend(r.urls)
