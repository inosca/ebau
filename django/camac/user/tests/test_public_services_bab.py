from datetime import date, timedelta

import pytest
from caluma.caluma_workflow.models import WorkItem
from django.urls import reverse
from rest_framework import status


@pytest.fixture
def so_services(service_factory, service):
    service_factory(
        name="my-subservice",
        service_group=service.service_group,
        service_parent=service,
    )

    for service_group, name in [
        # Restricted BaB services
        ("service-bab", "arp-bab"),
        # Restricted cantonal services
        ("service-cantonal", "arp"),
        ("service-cantonal", "arp-naturschutz"),
        ("service-cantonal", "arp-heimatschutz"),
        ("service-cantonal", "arp-nutzungsplanung"),
        ("service-cantonal", "arp-fuss-und-wanderwege"),
        ("service-cantonal", "denkmalpflege"),
        ("service-cantonal", "afu"),
        ("service-cantonal", "awjf"),
        ("service-cantonal", "gesundheitsamt"),
        ("service-cantonal", "avt"),
        ("service-cantonal", "alw"),
        # Unrestricted services
        ("municipality", "some-municipality"),
        ("service-cantonal", "some-cantonal-service"),
        ("service-extra-cantonal", "some-extra-cantonal-service"),
    ]:
        parent_service = service_factory(
            name=name,
            slug=name,
            service_group__name=service_group,
        )

        service_factory(
            name=f"{name}-subservice",
            service_group__name=service_group,
            service_parent=parent_service,
        )


@pytest.mark.parametrize(
    "service_group__name,conditions,expected_count",
    [
        pytest.param("municipality", ["authority"], 14, id="authority_bib"),
        pytest.param("municipality", ["authority", "bab"], 4, id="authority_bab"),
        pytest.param(
            "municipality",
            ["authority", "bab", "appeal"],
            5,
            id="authority_bab_with_appeal",
        ),
        pytest.param(
            "municipality",
            ["authority", "bab", "completed_publication"],
            5,
            id="authority_bab_with_completed_publication",
        ),
        pytest.param(
            "municipality",
            ["authority", "bab", "running_publication"],
            4,
            id="authority_bab_with_running_publication",
        ),
        pytest.param(
            "municipality",
            ["authority", "bab", "completed_publication", "running_publication"],
            4,
            id="authority_bab_with_completed_and_running_publication",
        ),
        pytest.param("service-cantonal", [], 14, id="service-cantonal"),
        pytest.param("service-extra-cantonal", [], 14, id="service-extra-cantonal"),
        pytest.param("municipality", [], 1, id="municipality"),
        pytest.param("service-bab", [], 14, id="service-bab"),
    ],
)
def test_so_distribution_services(
    db,
    admin_client,
    conditions,
    expected_count,
    mocker,
    service_factory,
    service,
    snapshot,
    so_appeal_settings,
    so_bab_settings,
    so_distribution_settings,
    so_instance,
    so_publication_settings,
    so_services,
    utils,
    work_item_factory,
):
    mocker.patch(
        "camac.instance.models.Instance.responsible_service",
        return_value=service if "authority" in conditions else service_factory(),
    )

    if "bab" in conditions:
        so_instance.case.meta["is-bab"] = True
        so_instance.case.save()

    if "appeal" in conditions:
        so_instance.case.meta["is-appeal"] = True
        so_instance.case.save()

    if "completed_publication" in conditions:
        work_item = work_item_factory(
            task_id=so_publication_settings["FILL_TASKS"][0],
            status=WorkItem.STATUS_COMPLETED,
            case=so_instance.case,
            meta={"is-published": True},
        )
        utils.add_answer(
            work_item.document,
            "publikation-ende",
            date.today() - timedelta(days=1),
        )

    if "running_publication" in conditions:
        work_item = work_item_factory(
            task_id=so_publication_settings["FILL_TASKS"][0],
            status=WorkItem.STATUS_COMPLETED,
            case=so_instance.case,
            meta={"is-published": True},
        )
        utils.add_answer(
            work_item.document,
            "publikation-ende",
            date.today() + timedelta(days=1),
        )
        utils.add_answer(
            work_item.document,
            "publikation-start",
            date.today() - timedelta(days=1),
        )

    response = admin_client.get(
        reverse("publicservice-list"),
        {
            "available_in_distribution_for_instance": so_instance.pk,
            "exclude_own_service": True,
        },
    )

    assert response.status_code == status.HTTP_200_OK

    data = response.json()["data"]

    assert len(data) == expected_count
    assert set([i["attributes"]["name"] for i in data]) == snapshot
