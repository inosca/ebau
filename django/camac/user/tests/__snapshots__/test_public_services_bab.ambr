# serializer version: 1
# name: test_so_distribution_services[authority_bab]
  set({
    'my-subservice',
    'some-cantonal-service',
    'some-extra-cantonal-service',
    'some-municipality',
  })
# ---
# name: test_so_distribution_services[authority_bab_with_appeal]
  set({
    'arp-bab',
    'my-subservice',
    'some-cantonal-service',
    'some-extra-cantonal-service',
    'some-municipality',
  })
# ---
# name: test_so_distribution_services[authority_bab_with_completed_and_running_publication]
  set({
    'my-subservice',
    'some-cantonal-service',
    'some-extra-cantonal-service',
    'some-municipality',
  })
# ---
# name: test_so_distribution_services[authority_bab_with_completed_publication]
  set({
    'arp-bab',
    'my-subservice',
    'some-cantonal-service',
    'some-extra-cantonal-service',
    'some-municipality',
  })
# ---
# name: test_so_distribution_services[authority_bab_with_running_publication]
  set({
    'my-subservice',
    'some-cantonal-service',
    'some-extra-cantonal-service',
    'some-municipality',
  })
# ---
# name: test_so_distribution_services[authority_bib]
  set({
    'afu',
    'alw',
    'arp-fuss-und-wanderwege',
    'arp-heimatschutz',
    'arp-naturschutz',
    'arp-nutzungsplanung',
    'avt',
    'awjf',
    'denkmalpflege',
    'gesundheitsamt',
    'my-subservice',
    'some-cantonal-service',
    'some-extra-cantonal-service',
    'some-municipality',
  })
# ---
# name: test_so_distribution_services[municipality]
  set({
    'my-subservice',
  })
# ---
# name: test_so_distribution_services[service-bab]
  set({
    'afu',
    'alw',
    'arp',
    'arp-fuss-und-wanderwege',
    'arp-heimatschutz',
    'arp-naturschutz',
    'arp-nutzungsplanung',
    'avt',
    'awjf',
    'denkmalpflege',
    'gesundheitsamt',
    'my-subservice',
    'some-cantonal-service',
    'some-extra-cantonal-service',
  })
# ---
# name: test_so_distribution_services[service-cantonal]
  set({
    'afu',
    'alw',
    'arp',
    'arp-fuss-und-wanderwege',
    'arp-heimatschutz',
    'arp-naturschutz',
    'arp-nutzungsplanung',
    'avt',
    'awjf',
    'denkmalpflege',
    'gesundheitsamt',
    'my-subservice',
    'some-cantonal-service',
    'some-extra-cantonal-service',
  })
# ---
# name: test_so_distribution_services[service-extra-cantonal]
  set({
    'afu',
    'alw',
    'arp',
    'arp-fuss-und-wanderwege',
    'arp-heimatschutz',
    'arp-naturschutz',
    'arp-nutzungsplanung',
    'avt',
    'awjf',
    'denkmalpflege',
    'gesundheitsamt',
    'my-subservice',
    'some-cantonal-service',
    'some-extra-cantonal-service',
  })
# ---
